File: string/20_minimum_window_substring.cpp
/*
    Minimum Window Substring - LeetCode #76
    
    Given two strings s and t, return the minimum window substring of s such that
    every character in t (including duplicates) is included in the window. If there
    is no such substring, return the empty string "".
    
    Example 1:
    Input: s = "ADOBECODEBANC", t = "ABC"
    Output: "BANC"
    Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.
    
    Example 2:
    Input: s = "a", t = "a"
    Output: "a"
    Explanation: The entire string s is the minimum window.
    
    Example 3:
    Input: s = "a", t = "aa"
    Output: ""
    Explanation: Both 'a's from t must be included in the window.
    
    Constraints:
    - 1 <= s.length, t.length <= 10^5
    - s and t consist of uppercase and lowercase English letters
    
    Time Complexity: O(n + m)
    Space Complexity: O(1) - fixed size of 52 or 128 for ASCII
*/

#include <iostream>
#include <string>
#include <unordered_map>
#include <climits>
using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty() || s.length() < t.length()) return "";
        
        unordered_map<char, int> tCount;
        for (char c : t) {
            tCount[c]++;
        }
        
        int required = tCount.size();
        int formed = 0;
        unordered_map<char, int> windowCount;
        
        int left = 0, right = 0;
        int minLen = INT_MAX;
        int minLeft = 0;
        
        while (right < s.length()) {
            // Add character from right
            char c = s[right];
            windowCount[c]++;
            
            if (tCount.count(c) && windowCount[c] == tCount[c]) {
                formed++;
            }
            
            // Try to contract window
            while (left <= right && formed == required) {
                // Update result
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minLeft = left;
                }
                
                // Remove from left
                char leftChar = s[left];
                windowCount[leftChar]--;
                
                if (tCount.count(leftChar) && windowCount[leftChar] < tCount[leftChar]) {
                    formed--;
                }
                
                left++;
            }
            
            right++;
        }
        
        return minLen == INT_MAX ? "" : s.substr(minLeft, minLen);
    }
};

int main() {
    Solution sol;
    
    cout << "Test 1: \"" << sol.minWindow("ADOBECODEBANC", "ABC") << "\"" << endl;  // "BANC"
    cout << "Test 2: \"" << sol.minWindow("a", "a") << "\"" << endl;                // "a"
    cout << "Test 3: \"" << sol.minWindow("a", "aa") << "\"" << endl;               // ""
    
    return 0;
}

File: string/20_minimum_window_substring.py
"""
    Minimum Window Substring - LeetCode #76
    
    Given two strings s and t, return the minimum window substring of s such that
    every character in t (including duplicates) is included in the window. If there
    is no such substring, return the empty string "".
    
    Example 1:
    Input: s = "ADOBECODEBANC", t = "ABC"
    Output: "BANC"
    Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.
    
    Example 2:
    Input: s = "a", t = "a"
    Output: "a"
    Explanation: The entire string s is the minimum window.
    
    Example 3:
    Input: s = "a", t = "aa"
    Output: ""
    Explanation: Both 'a's from t must be included in the window.
    
    Constraints:
    - 1 <= s.length, t.length <= 10^5
    - s and t consist of uppercase and lowercase English letters
    
    Time Complexity: O(n + m)
    Space Complexity: O(1) - fixed size of 52 or 128 for ASCII
"""

from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t or len(s) < len(t):
            return ""
        
        t_count = Counter(t)
        required = len(t_count)
        formed = 0
        window_count = {}
        
        left = 0
        right = 0
        min_len = float('inf')
        min_left = 0
        
        while right < len(s):
            # Add character from right
            char = s[right]
            window_count[char] = window_count.get(char, 0) + 1
            
            if char in t_count and window_count[char] == t_count[char]:
                formed += 1
            
            # Try to contract window
            while left <= right and formed == required:
                # Update result
                if right - left + 1 < min_len:
                    min_len = right - left + 1
                    min_left = left
                
                # Remove from left
                left_char = s[left]
                window_count[left_char] -= 1
                
                if left_char in t_count and window_count[left_char] < t_count[left_char]:
                    formed -= 1
                
                left += 1
            
            right += 1
        
        return "" if min_len == float('inf') else s[min_left:min_left + min_len]

if __name__ == "__main__":
    sol = Solution()
    
    print(f"Test 1: \"{sol.minWindow('ADOBECODEBANC', 'ABC')}\"")  # "BANC"
    print(f"Test 2: \"{sol.minWindow('a', 'a')}\"")                # "a"
    print(f"Test 3: \"{sol.minWindow('a', 'aa')}\"")               # ""
